### PR TITLE
Make ShadowCopyAnalyzerAssemblyLoader threadsafe

### PR DESCRIPTION
**Pull request by jasonmalinowski**
Multiple callers on multiple threads risked potentially overwriting the same files or having a mismatched Mutex which could have resulted in things being deleted underneath us. This is now properly safe.

There are two consumers of this type today: Visual Studio, which was currently calling this on the same thread the whole time (but that's changing) and VBCSCompiler, where it's a singleton and thus this might have been broken all along.

<details><summary>Ask Mode template</summary>

### Customer scenario

Using analyzers might have done bad things in VBCSCompiler, and will eventually do bad things once we land the free-threaded language service initialization work.

### Bugs this fixes

No bug, noticed as a part of other investigations.

### Workarounds, if any

None.

### Risk

Very low; just using standard patterns.

### Performance impact

None.

### Root cause analysis

Code was just never written to be thread-safe before.

### How was the bug found?

Feature work.

</details>

### CODE REVIEW

**Comment by sharwell** [comment]
:bulb: It looks like `CopyFileAndResources` and a bunch of methods below it can be marked `static`

**Comment by jasonmalinowski** [comment]
@dotnet/roslyn-compiler want to take a look since this affects your deliverables too?

**Comment by jasonmalinowski** [comment]
@dotnet-bot retest this please.

File: src/Compilers/Shared/ShadowCopyAnalyzerAssemblyLoader.cs [hunk 1]
```diff
@@ -49,6 +48,9 @@ public ShadowCopyAnalyzerAssemblyLoader(string baseDirectory = null)
                 _baseDirectory = Path.Combine(Path.GetTempPath(), "CodeAnalysis", "AnalyzerShadowCopies");
             }
 
+            _shadowCopyDirectoryAndMutex = new Lazy<(string directory, Mutex)>(
```
Comment (jcouv) [inline src/Compilers/Shared/ShadowCopyAnalyzerAssemblyLoader.cs]: >Lazy [](start = 47, length = 4)

What's the benefit for the `Lazy` here, compared to doing `_shadowCopyDirectoryAndMutex = CreateUniqueDirectoryForProcess();` in `LoadImpl(...)` method (where the `string` and `Mutex` where previously getting initialized)?

**Comment by jcouv** [comment]
@dotnet/roslyn-compiler for a second review. Thanks

File: src/Compilers/Shared/ShadowCopyAnalyzerAssemblyLoader.cs (hunk 1 above)
Comment (jasonmalinowski) [reply src/Compilers/Shared/ShadowCopyAnalyzerAssemblyLoader.cs]: LoadImpl might get called from multiple threads at once, but we want to have one directory and Mutex to cover it. The use of the Mutex here is so other processes can ask "is this directory still being used by some other process?" so we avoid deleting files out from underneath the owner. If you overwrite your previous directory/mutex, you're leaving those files open for collection by another process.

**Comment by jcouv** [comment]
@dotnet/roslyn-compiler for a second review. Thanks

**Comment by jinujoseph** [comment]
Approved for 15.8.Preview4 
cc @jaredpar  (approving as its been reviewed by compiler team)

**Pull request merged**